python/tests/xir/test_program.py

```diff
@@ -35,50 +26,40 @@ def test_includes(self):
     # Test declarations
     #####################
 
-    @pytest.mark.parametrize("name", ["rx", "CNOT", "a_gate"])
-    @pytest.mark.parametrize("num_params", [0, 1, 42])
-    @pytest.mark.parametrize("num_wires", [1, 42])
-    def test_gate_declaration(self, name, num_params, num_wires):
-        """Test serializing gate declarations"""
-        decl = GateDeclaration(name, num_params=num_params, num_wires=num_wires)
+    @pytest.mark.parametrize("params", [["a", "b"], [], ["theta"]])
+    @pytest.mark.parametrize("wires", [(0,), (0, 2, 1), ("a", "b", "c")])
+    @pytest.mark.parametrize("declaration_type", ["gate", "operator", "output"])
+    def test_declarations(self, params, wires, declaration_type):
+        """Test serializing gate, operation and output declarations"""
+        decl = Declaration("name", params, wires, declaration_type=declaration_type)
 
         irprog = XIRProgram()
-        irprog._declarations["gate"].append(decl)
+        irprog._declarations[declaration_type].append(decl)
         res = irprog.serialize()
-        assert res == f"gate {name}, {num_params}, {num_wires};"
 
-    @pytest.mark.parametrize("name", ["sin", "COS", "arc_tan"])
-    @pytest.mark.parametrize("num_params", [0, 1, 42])
-    def test_func_declaration(self, name, num_params):
-        """Test serializing function declarations"""
-        decl = FuncDeclaration(name, num_params=num_params)
+        if len(params) == 0:
+            params_str = ""
+        else:
+            params_str = "(" + ", ".join(params) + ")"
 
-        irprog = XIRProgram()
-        irprog._declarations["func"].append(decl)
-        res = irprog.serialize()
-        assert res == f"func {name}, {num_params};"
+        wires_str = ", ".join(str(w) for w in wires)
+        assert res == f"{declaration_type} name{params_str}[{wires_str}];"
 
-    @pytest.mark.parametrize("name", ["op", "OPERATOR", "op_42"])
-    @pytest.mark.parametrize("num_params", [0, 1, 42])
-    @pytest.mark.parametrize("num_wires", [1, 42])
-    def test_operator_declaration(self, name, num_params, num_wires):
-        """Test serializing operator declarations"""
-        decl = OperatorDeclaration(name, num_params=num_params, num_wires=num_wires)
+    @pytest.mark.parametrize("params", [["a", "b"], [], ["theta"]])
+    def test_func_declaration(self, params):
+        """Test serializing function declarations"""
+        decl = Declaration("name", params, (), declaration_type="function")
```

Comment:
I think this test would be cleaner if we made it as simple as possible:
```suggestion
    @pytest.mark.parametrize("params, want_res", [
            (["a", "b"], "function name(a, b);"),
            ([], "function name;"),
            (["theta"], "function name(theta);"),
        ],
    )
    def test_func_declaration(self, params, want_res):
        """Test serializing function declarations"""
        decl = Declaration("name", params, (), declaration_type="function")
        
        irprog = XIRProgram()
        irprog._declarations["func"].append(decl)

        have_res = irprog.serialize()
        assert have_res  == want_res
```
I would have suggested doing the same for `test_declarations()` above but the "cartesian product parameterization" could make this approach too verbose for that setting.

Reply:
Good suggestion! I actually changed `test_declarations` as well, since I think it's enough to test just a few examples here (all combinations aren't really strictly necessary I'd say).